### DISCUSSION TITLE
how should glsp gEdge make  a turn

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-glsp/glsp#1187] [Q&A]
for glsp gEdge,how can i make  a turn like below Identification section ,or for below diagram,which connection line should i use instead of gEdge(maybe i should custom?)
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/331c5fcd-27da-4212-8494-427f3d8c3aa4)

**Comment by rsoika** [comment]
It is not very clear to me what you want to achive. But for example you can implement your own Edge View. We do this in our project. Find the example code [here](https://github.com/imixs/open-bpmn/blob/master/open-bpmn.glsp-client/open-bpmn-glsp/src/bpmn-routing-views.tsx). You can also implement your own Router. But this is more complex and from your screenshot it seems not to be necessary.

**Reply by zrwsmd** [reply]
i mean the edge connection is not straight,but bended like below 
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/10f0f92d-ca7d-4e59-b657-4db2f734a941)

**Reply by rsoika** [reply]
This is called MANHATTAN routing. You can set the kind of routing in the edge properties.
See the example from our project: 


```java 
public class BPMNGEdgeBuilder extends AbstractGEdgeBuilder<BPMNGEdge, BPMNGEdgeBuilder> {

    private final String name;

    public BPMNGEdgeBuilder(final BPMNElementEdge edge) {
        super(edge.getType());
        this.name = edge.getName();
        this.id = edge.getId();
        this.addCssClass(type);
        this.addCssClass("bpmnedge");
    }

    @Override
    protected void setProperties(final BPMNGEdge edge) {
        super.setProperties(edge);
        edge.setName(name);
        // set the custom BPMN Router Kind
        edge.setRouterKind(GConstants.RouterKind.MANHATTAN);
        // set the label
        GLabel edgeLabel = new GLabelBuilder(ModelTypes.LABEL_HEADING) //
                .edgePlacement(new GEdgePlacementBuilder()//
                        .side(GConstants.EdgeSide.TOP)//
                        .position(0.5d) // center (50%)
                        .offset(3.0d) // 3px offset
                        .rotate(false) //
                        .build())//
                .id(id + "_bpmnlabel") //
                .text(edge.getName()).build();

        edge.getChildren().add(edgeLabel);
    }

    @Override
    protected BPMNGEdge instantiate() {
        return BpmnFactory.eINSTANCE.createBPMNGEdge();
    }

    @Override
    protected BPMNGEdgeBuilder self() {
        return this;
    }

}

```

**Reply by zrwsmd** [reply]
It does have that kind of effect，but connection seems random,the connection line  is irregular like below
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/fbd1503e-7d47-4664-a365-09433f1952bc)

**Reply by planger** [reply]
The Manhattan routing only does that if fixed routing points are defined in the GModel. Otherwise it goes the direct horizontal / vertical line from the source to the target. Do you have routing points in your GModel?

**Reply by zrwsmd** [reply]
if i want to have many routing points like below circle Identification section,how should i achieve it.actually.i am a little confused about how to use routing points,any example to recommend?official example(GRoutingHandleView]) a little confused
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/88c1b9fc-2033-4070-aeed-cf662767c4c6)

**Reply by planger** [reply]
I wouldn't implement the parts highlighted in your Ladder example with routing points.

I'd implement a diagram such as Ladder based on custom View implementations with ports.
Thus, I'd create a view implementation for rails (a vertical line) and position ports on them (where your red circles are). I'd implement also the contacts and coils as GModel elements with custom views that are made up of SVG lines and add ports left and right to them. Then I'd implement the rungs (horizontal lines) as edges that connect the ports of the rails to the ports of the contacts and coils.

This is roughly how the Ladder implementation by logi.cals has been implemented (plus they have automated layouting that computes the positions and sizes of rungs, contacts and coils in a matrix style):
https://eclipse.dev/glsp/gallery/#constrainedlayout

**Reply by planger** [reply]
Routing points are merely a mean to force the route of an edge to go through that point. It is usually only relevant when you have an edge and you want to steer the edge a certain route.

**Reply by zrwsmd** [reply]
did not clearly catch your meaning.for vertical line,i actually use a custom node extends GNode which width is 1 ,height is 400(for example),with Manhattan routing,it can achieve like below ,but when i add a new node,The previously added 2 nodes and their connections is always changing ,not fixed
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/18751d8f-2d18-4539-aff2-b34c64d26e02)

**Reply by planger** [reply]
This is very hard for me to debug remote without the code, but I'm pretty sure there are unwanted routing points in your edges. Your edges mustn't have any routing points and the port on the rail must be on the same height as the port (or anchor) on the rectangle. Then the lines would be straight horizontally.
So there is something going on in the GModel that you generate. I recommend to look into that.

**Reply by zrwsmd** [reply]
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/15ba0735-42c7-46d4-a2bf-0695788854b6)
when i drag a node  and create a connection( do save operation),it have routing point like below screnshot
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/e07bffb9-7dcb-4296-a72f-01994ab76d35)
but how to save this routing point so that next time when i reload model,i can render this routing point(reload and this routing point length is 0 at moment like below screenshot),but my previous saved edge is above screenshot
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/423cc951-9456-4ecf-8b4f-57e4ed814383)

**Reply by zrwsmd** [reply]
i find it computedBoundsAction  operation which return this routing point,but how this  routing point pass to gmodel factory
so that i can update gModel,i guess if this root(below screenshot)  call modelstate.updateRoot()  method to update root,it may suit my requirement
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/ab96ae74-44fa-443f-b878-6392b0cdec3b)

**Reply by planger** [reply]
I'm afraid I don't really understand what you want to achieve. Do you want to set the routing points on the server without having the client interfere? If yes, I suggest to configure the [polyline edge router](https://github.com/eclipse-sprotty/sprotty/blob/master/packages/sprotty/src/features/routing/polyline-edge-router.ts) on the client and add the routing points in your GModelFactory.
If you want your client to compute a Manhattan style routing (Manhattan routing will compute a route with horizontal and vertical lines from the source to the target point), then the server only sets the source and the target in the GModel and the client computes the route from source to target according to its strategy. If this routing strategy doesn't fit your needs, you can customize it on the client, too.

You can also mix this: the server sets certain fixed routing points and the client applies the routing strategy on top.

From your screenshots, it looks like you have the Manhattan routing configured on the client for this edge type. That means that this router will add routing points according to its routing strategy on the client. In the computed bounds action, the client will then report all computed sizes, positions and the routing point computed on the client to the server. This is why you seem to have two routing points in the GModel in the computed bounds action.

Just reading your last message, where you indicate you can address your requirements if you manipulate the GModel after the computed bounds action:
You can rebind the [ComputedBoundActionHandler](https://github.com/eclipse-glsp/glsp-server-node/blob/main/packages/server/src/common/features/layout/computed-bounds-action-handler.ts) with a your custom implementation to adjust the GModel after the client has communicated its computed sizes, positions, and routing points and change the GModel again before submitting it back to the client.

But as I said before, I'm not quite sure what your overall goal is.

**Reply by zrwsmd** [reply]
i drage two node and  connect them with Manhattan  edge,and i call save model method.it look like below  first screenshot
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/78ccc731-bcb4-4f4a-a9d8-62293c116756)
but when i call loadSourceModel method(reload it again) it like below which miss route point
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/d9280f87-dc22-4d9e-b1eb-bd6eda5dd3b3)


so i mean how can i save this routing point into gModel so that i can render it correctly
